### PR TITLE
Prevent renaming of Polymer 2 static properties

### PR DESCRIPTION
**Pull request by ChadKillingsworth**
Polymer 2 classes define static getters. When Polymer is used as an external library, these properties must not be renamed. Since the type system doesn't model static properties on a constructor currently, this pass adds `is` and `config` properties to `Function.prototype`. This will block renaming of any `is` or `config` property when defined on a function.

cc @justinfagnani

### CODE REVIEW

File: src/com/google/javascript/jscomp/DefaultPassConfig.java [hunk 1]
```diff
@@ -429,6 +429,10 @@ private void addOldTypeCheckerPasses(
       checks.add(generateExports);
     }
 
+    if (options.polymerPass && options.propertyRenaming != PropertyRenamingPolicy.OFF) {
```
Comment (justinfagnani) [inline src/com/google/javascript/jscomp/DefaultPassConfig.java]: does this need to be updated for the new flag names?

File: test/com/google/javascript/jscomp/PolymerProtectStaticPropertiesTest.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2008 The Closure Compiler Authors.
```
Comment (justinfagnani) [inline test/com/google/javascript/jscomp/PolymerProtectStaticPropertiesTest.java]: 2017?

File: src/com/google/javascript/jscomp/DefaultPassConfig.java (hunk 1 above)
Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/DefaultPassConfig.java]: Yes - just hadn't circled back since that landed.

**Comment by ChadKillingsworth** [comment]
Changes made - commits squashed.

**Comment by ChadKillingsworth** [comment]
Update to reflect the changes in Polymer/polymer/pull/4332

File: src/com/google/javascript/jscomp/PolymerProtectStaticProperties.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    compiler.reportCodeChange();
+  }
+
+  /** Lazily create a "new" externs root for undeclared variables. */
+  private Node getSynthesizedExternsRoot() {
+    return compiler.getSynthesizedExternsInput().getAstRoot(compiler);
+  }
+
+  /** Traverse the externs to find references to Polymer.Element. */
+  class FindPolymerExterns implements NodeTraversal.Callback {
```
Comment (justinfagnani) [inline src/com/google/javascript/jscomp/PolymerProtectStaticProperties.java]: Why doesn't the main PolymerPass need to find externs?

File: src/com/google/javascript/jscomp/PolymerProtectStaticProperties.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    this.compiler = compiler;
+  }
+
+  @Override
+  public void process(Node externs, Node root) {
+    FindPolymerExterns findPolymerExterns = new FindPolymerExterns();
+    NodeTraversal.traverseEs6(compiler, externs, findPolymerExterns);
+
+    if (findPolymerExterns.hasPolymerElementExterns()) {
+      addExtern("is");
```
Comment (justinfagnani) [inline src/com/google/javascript/jscomp/PolymerProtectStaticProperties.java]: We may add optional mixins that support `hostAttributes` and `listener` static properties. How should we handle those? The set of important statics is nearly as open as the set of instance methods now. Is there any way the _users_ of these properties can protect them or allow for renaming?

That seems to be what this is attempting: https://github.com/Polymer/polymer/blob/master/lib/mixins/element-mixin.html#L141

Is this not sufficient?

File: src/com/google/javascript/jscomp/PolymerProtectStaticProperties.java (hunk 3 above)
Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/PolymerProtectStaticProperties.java]: Currently, only Polymer 1 version passes is making use of the externs directly. Checking for the extern being present is already done by other passes in the compiler - namely you'll have type annotations for unknown symbols if they are missing.

Also, the long term plan was that Polymer 2 would be included in the compilation - not an extern.

File: src/com/google/javascript/jscomp/PolymerProtectStaticProperties.java (hunk 4 above)
Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/PolymerProtectStaticProperties.java]: No - it's not. With the class based style, the compiler does MUCH better at understanding the syntax and fully enables optimizations. I think the real answer is https://github.com/google/closure-compiler/issues/1693#issuecomment-303981793 but I don't know what it would take to land such a change at Google.

**Comment by justinfagnani** [comment]
Hey @ChadKillingsworth can we get a rebase? :)

**Comment by ChadKillingsworth** [comment]
@justinfagnani Done.

FYI This pass is a temporary workaround. I have a general plan with consensus of the compiler team on how to handle this in the general case for ES6 static properties and methods.

**Comment by justinfagnani** [comment]
cc @azakus 

Would this PR be merged, even temporarily?

**Comment by ChadKillingsworth** [comment]
Yes I would merge it temporarily. The changes needed for static properties might take just a bit for me to get worked out, then could take a while to get landed. After static properties are handled properly, we'll just remove this pass.

TL;DR - I have to ensure that the disambiguate and ambiguate properties passes are updated to account for static inheritance. Collapse Properties needs updated to not collapse static properties from ES6 classes.

**Comment by azakus** [comment]
@ChadKillingsworth is this PR still necessary after #2338?
